Sync shipping address with billing address when shipping address fields are disabled

mikejolley opened the pull request:
### Background

Taxes in WC are calculated based on _shipping address_.

During checkout, we conditionally show shipping address fields based on a "cartNeedsShipping" constant, which is derived from products being sold and shipping settings. If shipping fields are not shown, we usually just show billing fields only.

Whilst WooCommerce core does something similar on checkout, it keeps the shipping address synced with billing address so that taxes still update as needed.

In the case of blocks, when shipping address fields are hidden, we're not syncing or updating billing address via the API, which results in incorrect taxes being displayed (#3341) and taxes not being recalculated when the order is placed.

In fact, if you test this using the instructions in #3341 you'll notice that if you go ahead and place the order, the billing address is saved, but the shipping address uses whatever was in the session previously, or is left blank if checking out as a guest. The taxes are shown based on the initial shipping address as you entered checkout.

### Solution

To fix this we need to:

- Keep shipping and billing addresses synced in the client and server
- Send addresses to the server and return updated carts

This PR introduces `CustomerDataProvider ` to store billing and shipping addresses and pass down via context.

To sync data to the server, a new hook named `useCustomerData` stores the addresses and pushes to the API via a new `update-customer` endpoint. This replaces the `update-shipping` endpoint. This is debounced to prevent multiple requests in quick sucession.

Fixes #3341

### How to test the changes in this Pull Request:

**Case 1**
1. Create 2 tax rates; one for UK and one for US
2. Delete all shipping methods and rates so shipping is disabled and no fields are shown during checkout
3. Go to checkout. Enter a UK address. See the tax rate.
4. Change country field. See the tax rates update after the API request completes.
5. Place order. Check the order has the same shipping and billing address saved.

**Case 2**
1. In general settings, in the option for where you ship to, select 'disable shipping calculations'
2. Go to checkout. Enter a UK address. See the tax rate.
3. Change country field. See the tax rates update after the API request completes.
4. Place order. Check the order has the same shipping and billing address saved.

**Case 3**
1. In tax settings, choose to base taxes on the billing address.
2. Go to checkout. Enter a different billing and shipping address.
3. Toggle the "ship to billing" checkbox and see if the rate changes after doing so reliably.

In all cases you should see a network request to `update-customer` a short while after entering any billing or shipping data, or toggling the 'use shipping address' checkbox on checkout.

### Changelog

> Sync shipping address with billing address when shipping address fields are disabled, and update billing address on the server using a new `update-customer` REST API endpoint.

mikejolley committed:
Correct docblock description

mikejolley committed:
Sync shipping address changes with billing data

mikejolley committed:
Update inline documentation

senadir commented:
Instead of using the testing instructions above, I disabled shipping from `Settings -> General -> General Options -> Shipping locations` and I can't proceed, any change in Checkout would trigger an error `Shipping is disabled`

mikejolley commented:
@senadir Isn't that an unrelated issue? It's a bug but it's not what was logged in #3341

senadir commented:
![image](https://user-images.githubusercontent.com/6165348/97996004-8c188480-1de7-11eb-94dd-80d00717e67e.png)
![image](https://user-images.githubusercontent.com/6165348/97996021-920e6580-1de7-11eb-827f-2a6229ed8038.png)

When going over the testing instructions, I have two taxes for Alberta and British Columbia, switching between them doesn't update taxes, I can see the network request for `update-shipping` but taxes are still empty.

Using classic checkout, I can see the taxes being applied
![image](https://user-images.githubusercontent.com/6165348/97996560-342e4d80-1de8-11eb-8308-e8cb508579ce.png)

senadir commented:
@mikejolley Well that's how I filled the original issue (disabled shipping), there are several ways to disable shipping, this PR only fixes one of them, but the original issue would stay the same.

senadir commented:
If you want to handle the backend in a follow-up, that's okay, but should #3341 should stay open (and possibly reformated).

senadir commented:
The above issue (taxes not updating) happens when I have taxes based on billing address
![image](https://user-images.githubusercontent.com/6165348/97997173-faaa1200-1de8-11eb-9540-e6ff42c7ae7f.png)

mikejolley commented:
3 in 1 then :) yay

mikejolley committed:
Revert address sync because it fails when shipping is disabled explicitely

mikejolley committed:
Avoid loading shipping address from customer is shipping is disabled

mikejolley committed:
Rather than update order from the wc/store/checkout request, update the customer object

This is turn is synced to order, but also allows the cart calcultions to use the posted data. This means that taxes will be updated based on address data even if not displayed on the checkout.

mikejolley committed:
Add action that combines billing and shipping updates

mikejolley committed:
Add route for updating billing and shipping address

mikejolley committed:
Sync billing data to server on change

mikejolley commented:
@senadir This still needs work but can you take a quick peek? I think this is the way we need to solve it; send the billing data to the server when fields change, because there are cases where it's used for taxes (not just shipping).

mikejolley committed:
Shared constants for billing data

senadir commented:
Doing a functionality review before digging into code.

mikejolley committed:
Skip address update if missing country

senadir commented:
So the summary is:
- When shipping is disabled, this works fine ✅
- When shipping is enabled but there are no shipping methods, this works fine ✅
- When shipping is enabled and there are shipping methods:
    - If you have taxes calculated based on the **shipping address:**
        - Updating shipping form will update taxes ✅
        - Updating billing form would produce an error `No address country code was given` 🔴
        - Having `Use same address for billing` would update taxes but still produce the above error 🔴.   
    - If you have taxes calculated based on the **billing address:**
        - Updating billing form would produce an error `No address country code was given` 🔴
        - Having `Use same address for billing` would **not** update taxes but still produce the above error 🔴.   

I hope that made sense?

senadir commented:
Well I see that you added a38ef62 that should solve them, testing again.

mikejolley commented:
Thanks @senadir I noticed those issues too and resolved.

Do you think we should update the stores conditionally to avoid more api requests? It would be looking at quite a few core settings to determine if billing needs to be pushed to the server or not.

mikejolley committed:
Allow null values to skip formatting

mikejolley committed:
Add billing to cart schema

mikejolley committed:
Removed unwanted hooks from previous commit

mikejolley committed:
Decoding is handled in useStoreCart

mikejolley committed:
Remove hook

mikejolley committed:
Make shipping context hold state

mikejolley committed:
Make billing context hold state

mikejolley committed:
Add address processors

mikejolley commented on assets/js/base/context/cart-checkout/billing/index.js:
```diff
@@ -26,8 +46,15 @@ export const useBillingDataContext = () => {
+ */
 export const BillingDataProvider = ( { children } ) => {
-	const [ billingData, setBillingDataState ] = useState( DEFAULT_STATE );
+	const { billingAddress: initalBillingData } = useStoreCart();
```
Added to store cart and API for consistency with shipping_address

mikejolley commented on assets/js/base/context/cart-checkout/cart/index.js:
```diff
@@ -22,6 +23,7 @@ export const CartProvider = ( { children, redirectUrl } ) => {
 			<ShippingDataProvider>
 				<PaymentMethodDataProvider>
 					{ children }
+					<CartProcessor />
```
Similar pattern to checkout in that it consumes contexts further up the tree. In this case it listens for shipping and billing address updates and dispatches updates to the API on change.

mikejolley committed:
Cart does not have billing

mikejolley commented on assets/js/base/context/cart-checkout/cart/processor/index.js:
```diff
@@ -0,0 +1,56 @@
+	// shipping addresses to the server and get an updated cart in a single request.
+	useEffect( () => {
+		if (
+			shouldUpdateAddressStore(
```
If the address is changed, and has a `country`, we dispatch the update to the server. This is instead of the `useShippingAddress` hook which was only consumed by these contexts.

mikejolley commented on assets/js/base/context/cart-checkout/checkout/processor/index.js:
```diff
@@ -18,33 +19,13 @@ import {
- * @return {PaymentDataItem[]} Returns the payment data as an array of
- *                             PaymentDataItem objects.
- */
-const preparePaymentData = ( paymentData, shouldSave, activePaymentMethod ) => {
```
Moved to utils to keep this file smaller.

mikejolley commented on assets/js/base/context/cart-checkout/checkout/processor/index.js:
```diff
@@ -262,6 +206,94 @@ const CheckoutProcessor = () => {
+		currentRedirectUrl.current = redirectUrl;
+	}, [ billingData, shippingAddress, redirectUrl ] );
+
+	// When the billing or shipping address changes we need to push the changes to the server to
```
This monitors for address changes to both billing and shipping so they are updated together in a single request.

mikejolley commented on assets/js/base/context/cart-checkout/checkout/processor/index.js:
```diff
@@ -262,6 +206,94 @@ const CheckoutProcessor = () => {
+				billing_address: {
+					...currentBillingData.current,
+					phone: currentBillingData.current.phone || null,
+					email: currentBillingData.current.email || null,
```
Using null prevents API validation for valid email addresses.

mikejolley commented on src/StoreApi/Routes/Checkout.php:
```diff
@@ -123,8 +123,10 @@ protected function get_route_response( \WP_REST_Request $request ) {
 	 * @return \WP_REST_Response
```
The changes within this class ensure that totals reflect customer data posted via the form, so if the API updates don't run on the checkout itself, the correct taxes etc are still calculated.

mikejolley committed:
Update tests, remove some unrelated changes affecting the diff

mikejolley committed:
Revert "Update inline documentation"

This reverts commit 0393f49316de3152c6dcf6fda1192c06a74f1b55.

mikejolley committed:
Make shippingRatesAreResolving conditonal based on API request

mikejolley committed:
Shared address processor in cart and checkout

mikejolley committed:
Rename REST endpoint

nerrad reviewed (commented):
I haven't done an exhaustive review of this. In general I think you've identified a couple flaws with the current architectural implementation of Contexts in the code:

- Woo Core has a tight coupling between shipping and billing address with regards to representing _customer data_. 
- Woo Blocks was trying to keep these as discrete data sets that would get synced in key places.

What we've realized is that billing data and shipping data must _always_ be in sync regardless of the state of the UI/UX. 

We've already implemented a number of fixes due to this flawed architectural assumption (which is largely my fault) so I think we need to rethink this a bit.

Should we align around the new `CartUpdateAddress` route (and endpoint) for all address updates or is there still need for the separate endpoints? I'm wary about having more than one endpoint for updating discrete data (and there seems to be overlap between `CartUpdateAddress`, and `CartUpdateShipping`. `Routes\Checkout` also seems to have some overlap in that it updates addresses too.). It'd be nice to have a canonical endpoint for anything customer address related server side that all requests go through (which removes the guess work around what address endpoint to use).

Also, rather than an `AddressProcessor` what if we synchronize around this new endpoint for anything "customer address" related (which includes shipping and billing address data) and create a context specifically for _customer_data_? This would replace the Billing Data context and we'd then keep the shipping context just for handling the various updates related to shipping rates etc from shipping address updates.

> Note: I'm flexible on naming, `<CustomerAddressProvider />` would align more with the endpoint name primarily implemented in it (via `wp.data` store), but I'm also wondering if for extensibility it'd be good to make it a bit more generic and use it to represent anything related to customer data?

So the tree would be (for checkout):

```js
<CheckoutStateProvider>
    <CustomerDataProvider>
      <ShippingDataProvider>
        <PaymentMethodDataProvider>
          {children}
          <CheckoutProcessor/>
        </PaymentMethodDataProvider>
      </ShippingDataProvider>
    </CustomerDataProvider>
</CheckoutStateProvider>
```

The concern I have with the current approach is it appears to increase the complexity of the code (both in terms of number of endpoints and locations where address information is updated/retrieved) a bit more than I think is necessary and feels like another bandaid for the larger architectural problem (a clever bandaid at that 👍 ). I think what we've identified here is a need to keep customer data together (out of necessity with the WooCommerce data model), instead of splitting it into discrete data sets.

I realize, this would result in a more significant refactor. For payment methods at least, I think we'll still be able to preserve much of the public api exposed via `usePaymentMethodInterface` so that shouldn't be affected too much (beyond changing some of the data sources). However, I think we'll be able to simplify a lot of logic and flow as a result of the refactor too (and hopefully eliminate some code).

mikejolley committed:
CustomerDataProvider and hook

mikejolley committed:
Update shipping address type defs

mikejolley committed:
Rename customer address endpoint, and remove update-shipping

mikejolley committed:
Update tests

mikejolley committed:
Fix tests by restoring country validation

mikejolley commented:
This is ready for another review.

mikejolley committed:
typo

nerrad commented on assets/js/base/context/cart-checkout/checkout/processor/index.js:
```diff
@@ -66,8 +44,8 @@ const CheckoutProcessor = () => {
-	const { shippingAddress, shippingErrorStatus } = useShippingDataContext();
-	const { billingData } = useBillingDataContext();
+	const { shippingErrorStatus } = useShippingDataContext();
+	const { billingData, shippingAddress } = useCustomerDataContext();
```
Gotta say, I like this import better. Seems to make more sense to have these values together.

nerrad commented on assets/js/base/context/cart-checkout/checkout/processor/utils.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Utility function for preparing payment data for the request.
+ *
+ * @param {Object}  paymentData          Arbitrary payment data provided by the payment method.
+ * @param {boolean} shouldSave           Whether to save the payment method info to user account.
+ * @param {Object}  activePaymentMethod  The current active payment method.
```
Ideally we'd have `paymentData` and `activePaymentMethod` defined as typedefs so we are explicit about what is passed through here.

mikejolley replied:
As this is unrelated to this PR I won't add it here, but I did notice this already has a todo: ```@todo do typedefs for the payment event state.``` in payment-methods/constants.js

nerrad commented on assets/js/base/hooks/cart/use-store-cart.js:
```diff
@@ -9,6 +9,35 @@ import { useEditorContext } from '@woocommerce/base-context';
+	state: '',
+	postcode: '',
+	country: '',
+};
```
```suggestion
const defaultShippingAddress = {
	first_name: '',
	last_name: '',
	company: '',
	address_1: '',
	address_2: '',
	city: '',
	state: '',
	postcode: '',
	country: '',
};

const defaultBillingAddress = {
	...defaultShippingAddress,
	email: '',
	phone: '',
};
```

nerrad commented on assets/js/base/hooks/customer/use-customer-data.js:
```diff
@@ -0,0 +1,102 @@
+		shippingAddress: initialShippingAddress,
+	} );
+	const previousCustomerData = useRef( customerData );
+	const [ debouncedCustomerData ] = useDebounce( customerData, 400 );
```
It looks like you can simplify things a bit by [passing in the equality function](https://github.com/xnimorz/use-debounce/blob/6f86d15fb4b112e9a9fb86ad90f29ca98f5332e2/src/useDebounce.ts#L11) to `useDebounce`. This would allow you to remove the comparison logic for `shouldUpdateBilling` and `shouldUpdateShipping` out of the below `useEffect` and also eliminate the extra `useRef` in the `useCustomerData` hook.

So something like:

```js
const [ debouncedCustomerData ] = useDebounce( 
  customerData,
  400,
  { 
    equalityFn: ( prevData, newData ) => {
      return shouldUpdateAddressStore( prevData.billingData, newData.billingData ) || shouldUpdateAddressStore( prevData.shippingAddress, newData.shippingAddress )
    }
  }
)
```

I realize the one downside to this approach is you won't know explicitly whether shipping address or billing address changed - but I'm not sure I understand why we care? Can't we just always send both the current debounced shipping address value and debounced billing address value in a request when there's a change in either (and let the server handle what addresses actually had a change)?

mikejolley replied:
Yes but in the action that we dispatch we're setting some things to true when shipping rates are resolving. If we are only updating billing address that logic isn't needed. That could impact performance?

nerrad replied:
Hmm, I see, and `shippingRatesAreResolving` will in turn mean the ui/ux for shipping rates will be showing a loading indicator everytime _billing address_ is changed right?

So it does look it's better to keep things as they are for now.

mikejolley replied:
I went ahead with this feedback and replaced the shipping rate resolving code with customer data resolving.

nerrad reviewed (commented):
In general looks great Mike! I like the refactor and I think it makes the logic more straightforward having Shipping and Billing data live together.

I do have some feedback for the `useCustomerData` hook that might be applicable.

I have not tested yet, but am happy to assist with that once the feedback is responded to/addressed. I'm mostly wanting to make sure express payment methods still work with the changes (they use the shipping address apis for getting rates etc). From what I've seen in the code I don't think our changes here should impact that at all but still good to verify.

mikejolley committed:
Update assets/js/base/hooks/cart/use-store-cart.js

Co-authored-by: Darren Ethier <darren@roughsmootheng.in>

mikejolley committed:
Simplify debounce and request handling

mikejolley committed:
Remove state from address sync

This will mean billing is "forgotten" if using the checbox, but this greatly simplifies logic.

mikejolley committed:
Rename shipping rates loading to customer data loading

mikejolley committed:
Sync based on useStoreCart data

mikejolley committed:
Made cart API less strict on addresses

mikejolley committed:
Fix useCheckoutAddress sync

mikejolley commented:
I think this is working now — tested with Stripe Chrome Pay. Would appreciate some help testing various cases :)

nerrad commented on assets/js/base/hooks/checkout/use-checkout-address.js:
```diff
@@ -46,55 +43,54 @@ export const useCheckoutAddress = () => {
-		setBillingData,
-		setShippingAddress,
-	] );
+	const currentShippingAsBilling = useRef( shippingAsBilling );
```
I'm not sure about this variable name. Is it really current as of the execution time? We typically use the `previous` prefix for `useRef` references (which is how it appears to be used in the logic here).

mikejolley replied:
This is current as in the context of `useCheckoutAddress` — if you see the return value from this hook it uses the `currentShippingAsBilling ` ref rather than `shippingAsBilling `. The reason I had to do it this way was because our billing address fields contain debounce code on the values, so if you return `shippingAsBilling `, those fields are rendered with the old values before this hook has had a chance to sync billing data.

I will add a comment to that effect if it helps.

nerrad commented on assets/js/base/hooks/checkout/use-checkout-address.js:
```diff
@@ -46,55 +43,54 @@ export const useCheckoutAddress = () => {
+				previousBillingData.current = billingData;
+				setBillingData( shippingAddress );
+			} else {
+				setBillingData( {
+					...previousBillingData.current,
+					email: undefined,
+					phone: undefined,
+				} );
```
Won't this clear out previous set email and phone number if the `shippingAsBilling` checkbox is unchecked? Do we want that behaviour?

mikejolley replied:
I dont think so--we merge billing data, so existing emails are preserved. I'll double check tommorow.

mikejolley replied:
Confirmed this is a merge update.

nerrad commented on assets/js/base/hooks/checkout/use-checkout-address.js:
```diff
@@ -46,55 +43,54 @@ export const useCheckoutAddress = () => {
-	const setPhone = ( value ) => void setBillingFields( { phone: value } );
+	// This syncs billing data with shipping data if the checkbox is enabled.
+	useEffect( () => {
+		if ( currentShippingAsBilling.current !== shippingAsBilling ) {
```
So just for clarity sake - I was originally wondering why `shippingAsBilling` was being tracked as a ref but it looks like why it's implemented this way is because:

- the `billingData` dependency could trigger this effect frequently (because it's an object that changes).
- tracking previous state of `shippingAsBilling` allows for escaping the execution of this logic and only doing so when the `shippingAsBilling` state changes.

The potential problem here is that we still want billing data to update when shipping data changes when `shippingAsBilling` is true but this problem is resolved  by the `setShippingFields` callback.

Is the above correct?

mikejolley replied:
Right. This effect basically only handles the toggle event; when the checkbox is toggled we need to immediately update billing address to reflect this change. I will add a comment.

nerrad commented on assets/js/base/hooks/customer/use-customer-data.js:
```diff
@@ -19,73 +20,94 @@ export const useCustomerData = () => {
 	const {
-		billingAddress: initialBillingData,
-		shippingAddress: initialShippingAddress,
+		billingAddress: cartBillingData,
```
I'm getting a typedef warning here for property `billingAddress` not existing for type `StoreCart`.

mikejolley replied:
I think you need to refresh your IDE; it's definitely defined.

nerrad replied:
ahh okay, maybe it was due to me switching branches or something then.

nerrad commented on assets/js/base/hooks/customer/use-customer-data.js:
```diff
@@ -19,73 +20,94 @@ export const useCustomerData = () => {
-				requestData.shipping_address =
-					debouncedCustomerData.shippingAddress;
-			}
+	useEffect( () => {
+		if (
+			! isShallowEqual(
+				currentShippingAddress.current,
+				cartShippingAddress
+			)
+		) {
+			currentShippingAddress.current = cartShippingAddress;
+		}
+	}, [ cartShippingAddress ] );
```
Could the `currentShippingAddress` and `currentBillingAddress` still be in a single ref object and merge these two above effects so you'd end up with something like this?

```js
useEffect( () => {
  if (
    ! isShallowEqual(
      currentCustomerData.current.shippingAddress,
      cartShippingAddress
    )
  ) {
    currentCustomerData.current.shippingAddress = cartShippingAddress;
  }
  if (
    ! isShallowEqual(
      currentCustomerData.current.billingData,
      cartBillingData
    )
  ) {
    currentCustomerData.current.billingData = cartBillingData;
  }
}, [ cartShippingAddress, cartBillingData ] );
```

mikejolley replied:
I'm going to apply this feedback, however, I'm not sure what is really best practice or more performant. The above useEffect will be triggered more often than just the billing or shipping one, so maybe `isShallowEqual` will run more often? On the otherhand, this would make it more consistent with other data in this hook so doing it 👍🏻

nerrad commented on assets/js/base/hooks/customer/use-customer-data.js:
```diff
@@ -19,73 +20,94 @@ export const useCustomerData = () => {
+	useEffect( () => {
+		if (
+			! (
+				shouldUpdateAddressStore(
+					currentBillingData.current,
+					debouncedCustomerData.billingData
+				) ||
+				shouldUpdateAddressStore(
+					currentShippingAddress.current,
+					debouncedCustomerData.shippingAddress
+				)
+			)
```
Is this still needed if the `debouncedCustomerData` is already doing the equality check? Could you just rely on only updating customer data if `debouncedCustomerData` changes (which would be a dependency change)?

mikejolley replied:
Yes this is needed. 

`debouncedCustomerData ` stores all customer data debounced, using shallow equals. 

`shouldUpdateAddressStore` on the other-hand only looks at critical address fields (country, state, postcode) which need pushing to the server. 

Removing this would result in more unnecessary server requests.

nerrad commented on src/StoreApi/Routes/CartUpdateCustomer.php:
```diff
@@ -135,24 +137,17 @@ protected function get_route_post_response( \WP_REST_Request $request ) {
-				),
-				400
-			);
+			return [];
```
Can you provide some more explanation on why this was changed (I see the phpdoc but it's still not clear to me why this was changed from an exception to returning an empty array).

mikejolley replied:
Express payments don't always return a country code I found in testing. Additionally (and similarly) sometimes a billing email is provided but not the full address. This resulted in error 400.

At cart level, having a full complete address is not that important. We just want to track as much data as possible up until an order is created. This matches session behaviour in core. 

I'll update the doc block.

nerrad reviewed (commented):
Changes are looking good, I still have some feedback though. Also I haven't tested yet because the branch is out of date with master (and there's some conflicts with rebasing). Ping me when it's rebased and ready for testing and I'll take a look.

mikejolley committed:
Add note on currentShippingAsBilling

mikejolley committed:
Use incoming isCart

mikejolley committed:
Add more detailed inline comment for shippingAsBilling toggle event

mikejolley committed:
Combine customer billing and shipping ref

mikejolley committed:
Update address docblock

mikejolley committed:
Error handling in pluckAddress

mikejolley committed:
Fix cart response after rebase

mikejolley commented:
Feedback is addressed. Waiting on automated tests.

mikejolley committed:
Update customer tests

nerrad commented on src/StoreApi/Routes/CartUpdateCustomer.php:
```diff
@@ -145,7 +145,8 @@ protected function get_route_post_response( \WP_REST_Request $request ) {
 	 */
 	protected function prepare_address_fields( $address, $allowed_countries ) {
-		// Addresses require a country, otherwise are not valid--return an empty address.
+		// Addresses require a country, otherwise they are not valid. This return an empty address. There is no address
```
```suggestion
		// Addresses require a country, otherwise they are not valid. This returns an empty address. There is no address
```

nerrad reviewed (approved):
Thanks for addressing the feedback. I didn't get a chance to test but code looks good (aside from a minor typo that could be fixed). Pre-approving, your judgment call if you'd like someone else to do a confidence check on testing the changes.

mikejolley committed:
Update src/StoreApi/Routes/CartUpdateCustomer.php

Co-authored-by: Darren Ethier <darren@roughsmootheng.in>
